pkg/deploy/devfile_registry.go

```diff
@@ -27,49 +27,90 @@ type DevFileRegistryConfigMap struct {
 }
 
 const (
-	DevfileRegistry = "devfile-registry"
+	DevfileRegistry              = "devfile-registry"
+	devfileRegistryGatewayConfig = "devfile-registry-gtw"
 )
 
 /**
  * Create devfile registry resources unless an external registry is used.
  */
-func SyncDevfileRegistryToCluster(deployContext *DeployContext) (bool, error) {
+func SyncDevfileRegistryToCluster(deployContext *DeployContext, cheHost string) (bool, error) {
 	devfileRegistryURL := deployContext.CheCluster.Spec.Server.DevfileRegistryUrl
 	if !deployContext.CheCluster.Spec.Server.ExternalDevfileRegistry {
 		var host string
+		exposureStrategy := util.GetServerExposureStrategy(deployContext.CheCluster, DefaultServerExposureStrategy)
+		singleHostExposureType := GetSingleHostExposureType(deployContext.CheCluster)
+		useGateway := exposureStrategy == "single-host" && (util.IsOpenShift || singleHostExposureType == "gateway")
 		if !util.IsOpenShift {
```

Comment:
I feel like this could be simplified and extracted to functions with something like
```
if isOpenshift {
  if exposureStrategy == "multi-host" {
    exposeOpenshiftMultihost(...)
  } else if exposureStrategy == "single-host" {
    exposeOpenshiftSinglehost(...) 
  }
} else {
  if multi-host {
    exposeKubeMulti()
  } else if single-host {
 ......
```
wdyt?

Reply:
the same applies for plugin-registry and keycloak.